internal/controller/k0smotron.io/k0smotroncluster_pvc.go

```diff
@@ -0,0 +1,122 @@
+	}
+
+	return nil
+}
+
+func (r *ClusterReconciler) reconcileControlPlanePVC(ctx context.Context, kmc km.Cluster) error {
+	// Do nothing if the persistence type is not PVC
+	if kmc.Spec.Persistence.Type != "pvc" {
+		return nil
+	}
+
+	if kmc.Spec.Persistence.PersistentVolumeClaim.Name == "" {
+		kmc.Spec.Persistence.PersistentVolumeClaim.Name = kmc.GetVolumeName()
+	}
+
+	return r.resizeStatefulSetAndPVC(ctx, kmc, *kmc.Spec.Persistence.PersistentVolumeClaim.Spec.Resources.Requests.Storage(), kmc.Spec.Replicas, kmc.GetStatefulSetName(), kmc.Spec.Persistence.PersistentVolumeClaim.Name)
+}
+
+func (r *ClusterReconciler) reconcileEtcdPVC(ctx context.Context, kmc km.Cluster) error {
+	return r.resizeStatefulSetAndPVC(ctx, kmc, kmc.Spec.Etcd.Persistence.Size, calculateDesiredReplicas(&kmc), kmc.GetEtcdStatefulSetName(), "etcd-data")
+}
+
+func (r *ClusterReconciler) resizeStatefulSetAndPVC(ctx context.Context, kmc km.Cluster, desiredStorageSize resource.Quantity, replicas int32, stsName, vctName string) error {
+	var sts appsv1.StatefulSet
+	err := r.Get(ctx, client.ObjectKey{Namespace: kmc.Namespace, Name: stsName}, &sts)
+	if err != nil {
+		// Do nothing if StatefulSet does not exist yet
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to get statefulset: %w", err)
+	}
+
+	if desiredStorageSize.IsZero() ||
+		len(sts.Spec.VolumeClaimTemplates) == 0 ||
+		desiredStorageSize.Cmp(*sts.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()) == 0 {
+		return nil
+	}
+
+	var allowExpansion *bool
+	for i := 0; i < int(replicas); i++ {
+		var pvc corev1.PersistentVolumeClaim
+
+		name := fmt.Sprintf("%s-%s-%d", vctName, stsName, i)
+		err := r.Get(ctx, client.ObjectKey{Namespace: kmc.Namespace, Name: name}, &pvc)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// Do nothing if PVC does not exist yet
+				return nil
+			}
+			return fmt.Errorf("failed to get PVC %s: %w", name, err)
+		}
+
+		if allowExpansion == nil {
+			var sc storagev1.StorageClass
+			err = r.Get(ctx, client.ObjectKey{Name: *pvc.Spec.StorageClassName}, &sc)
+			if err != nil {
+				return fmt.Errorf("failed to get StorageClass %s: %w", *pvc.Spec.StorageClassName, err)
+			}
+			allowExpansion = sc.AllowVolumeExpansion
+		}
+
+		if allowExpansion != nil && *allowExpansion {
+			pvc.Spec.Resources.Requests[corev1.ResourceStorage] = desiredStorageSize
+			err = r.Update(ctx, &pvc)
+			if err != nil {
+				return fmt.Errorf("failed to update PVC %s: %w", pvc.Name, err)
+			}
+
+			// Remove pod to trigger file system resize
+			err = r.Delete(ctx, &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("%s-%d", stsName, i),
+					Namespace: kmc.Namespace,
+				},
+			}, &client.DeleteOptions{})
+
+			if err != nil {
+				return fmt.Errorf("failed to delete pod '%s' for resizing: %w", fmt.Sprintf("%s-%d", stsName, i), err)
+			}
+		} else {
+			// Do not check other PVCs if expansion is not allowed
```

Comment:
hmm, maybe we should somehow tell user this? I see two options, update the reconcile state and/or create an event

Reply:
Done